typed-protocols/src/Network/TypedProtocol/Core.hs

```diff
@@ -166,6 +170,70 @@ module Network.TypedProtocol.Core (
 -- each protocol that this property is not violated. It also provides utilities
 -- helpful for testing protocols.
 
+-- $lemmas
+--
+-- The 'connect' and 'connectPipelined' proofs rely on lemmas about the
+-- protocol. Specifically they rely on the property that each protocol state
+-- is labelled with the agency of one peer or the other, or neither, but never
+-- both. Or to put it another way, the protocol states should be partitioned
+-- into those with agency for one peer, or the other or neither.
+--
+-- The way the labelling is encoded does not automatically enforce this
+-- property. It is technically possible to set up the labelling for a protocol
+-- so that one state is labelled as having both peers with agency, or declaring
+-- simultaneously that one peer has agency and that neither peer has agency
+-- in a particular state.
+--
+-- So the overall proofs rely on lemmas that say that the labelling has been
+-- done correctly. This type bundles up those three lemmas.
+--
+-- Specifically proofs that it is impossible for a protocol state to have:
+--
+-- * client having agency and server having agency
+-- * client having agency and nobody having agency
+-- * server having agency and nobody having agency
+--
+-- These lemmas are structured as proofs by contradiction, e.g. stating
+-- \"if the client and the server have agency for this state then it leads to
+-- contradiction\". Contradiction is represented as the 'Void' type that has
+-- no values except ⊥.
+--
+-- For example for the ping\/pong protocol, it has three states, and if we set
+-- up the labelling correctly we have:
+--
+-- > data PingPong where
+-- >   StIdle :: PingPong
+-- >   StBusy :: PingPong
+-- >   StDone :: PingPong
+-- >
+-- > data ClientHasAgency st where
+-- >   TokIdle :: ClientHasAgency StIdle
+-- >
+-- > data ServerHasAgency st where
+-- >   TokBusy :: ServerHasAgency StBusy
+-- >
+-- > data NobodyHasAgency st where
+-- >   TokDone :: NobodyHasAgency StDone
+--
+-- So now we can prove that if the client has agency for a state then there
+-- are no cases in which the server has agency.
+--
+-- > proofByContradiction_ClientAndServerHaveAgency TokIdle tok =
+-- >   case tok of {}
+--
+-- For this protocol there is only one state in which the client has agency,
+-- the idle state. By pattern matching on the state token for the server
+-- agency we can list all the cases in which the server also has agency for
+-- the idle state. There are of course none of these so we give the empty
+-- set of patterns. GHC can check that we are indeed correct about this.
+-- This also requires the @EmptyCase@ language extension.
+--
+-- To get this completeness checking it is important to compile modules
+-- containing these lemmas with @-Wincomplete-patterns@, which is implied by
+-- @-Wall@.
```

Comment:
Good explanation! Should also put in a derived lemma which switches the client and server parts.

```Haskell
proofByContradiction_ServerExclusive :: ServerHasAgency st -> ClientHasAgency st -> Void
proofByContradiction_ServerExclusive = \server -> \client -> 
  proofByContradiction_ClientAndServerHaveAgency client server
```

Reply:
Thanks, that's what I added previously, based on your critique.

Not a bad idea, adding a flipped one. Currently we just use it flipped in the `connect` proofs.